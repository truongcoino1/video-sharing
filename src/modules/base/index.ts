export * from './types/api.d'
export * from './services'
export * from './hooks'