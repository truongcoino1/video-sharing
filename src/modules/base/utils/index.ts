export function flushPromises(timeout = 0): Promise<unknown> {
  return new Promise(function (resolve) {
    setTimeout(resolve, timeout);
  });
}

export const uniqueArrayById = (array: any[]) => {
  return array.filter((v, i, a) => a.findIndex((t) => t.id === v.id) === i && !!v.id);
};
