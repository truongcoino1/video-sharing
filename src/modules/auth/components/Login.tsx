import React, { memo, useCallback, useMemo } from "react";
import { Button, useInput } from "@/modules/base";
import { useAuth } from "../hooks/useAuth";

const Login = () => {
  const { loading, login, error } = useAuth();
  const { value: email, handleOnChange: handleSetEmail } = useInput("");
  const { value: password, handleOnChange: handleSetPassword } = useInput("");

  const handleLogin = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      login(email || "", password || "");
    },
    [email, password]
  );

  const isDisabled = useMemo(() => {
    return !email || !password;
  }, [email, password]);

  return (
    <div className="login-container">
      <form
        className="login-form flex-col flex lg:flex-row"
        onSubmit={handleLogin}
      >
        <input
          autoFocus
          name="email"
          type="email"
          placeholder="Email"
          aria-label="email"
          value={email}
          className="lg:mr-8 outline-none h-[40px] px-8 mt-12 lg:mt-0 border rounded-medium border-gray-300"
          onChange={handleSetEmail}
        />
        <input
          name="password"
          type="password"
          placeholder="Password"
          aria-label="password"
          value={password}
          className="lg:mr-8 h-[40px] outline-none px-8 border mt-12 lg:mt-0 rounded-medium border-gray-300"
          onChange={handleSetPassword}
        />
        <Button isLoading={loading} type="submit" isDisable={isDisabled} label="Login / Register" className="px-8 lg:mt-0 h-[40px] mt-12 rounded-medium" />
      </form>
    </div>
  );
};

export default memo(Login);
