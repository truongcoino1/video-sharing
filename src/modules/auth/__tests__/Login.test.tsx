import { render, cleanup, fireEvent } from "@testing-library/react";
import Login from "../components/Login";
import {flushPromises} from '@/modules/base';
import "@/modules/base/utils/jest.mock";

const login = jest.fn();
jest.mock("../hooks/useAuth", () => {
  const useAuth = () => {
    return {
      login: login,
    };
  };
  return {
    useAuth,
  };
});

describe("Login", () => {
  afterEach(cleanup);

  it("Login render without crash", () => {
    const result = render(<Login />);
    const container = result.container;
    const elm = container.querySelector(".login-container");
    const inputEmail = container.querySelector('input[type="email"]');
    const inputPassword = container.querySelector('input[type="password"]');
    const btnSubmit = container.querySelector('button[type="submit"]');
    expect(elm).toBeInTheDocument();
    expect(inputEmail).toBeInTheDocument();
    expect(inputPassword).toBeInTheDocument();
    expect(btnSubmit).toBeInTheDocument();
  });

  it("Button Login/Register disabled when email and password empty", () => {
    const result = render(<Login />);
    const btnSubmit = result.container.querySelector('button[type="submit"]');
    expect(btnSubmit).toBeDisabled();
  });

  const setUp = () => {
    const result = render(<Login />);
    const container = result.container;
    const inputEmail = container.querySelector(
      'input[type="email"]'
    ) as Element;
    const inputPassword = container.querySelector(
      'input[type="password"]'
    ) as Element;
    const btnSubmit = container.querySelector(
      'button[type="submit"]'
    ) as HTMLButtonElement;
    const form = container.querySelector("form") as Element;
    return {
      inputEmail,
      inputPassword,
      btnSubmit,
      form,
    };
  };

  it("Button Login/Register disabled when email empty and password not empty", () => {
    const { inputPassword, btnSubmit } = setUp();
    fireEvent.change(inputPassword, { target: { value: "test_password" } });
    expect(btnSubmit).toBeDisabled();
  });

  it("Button Login/Register disabled when email not empty and password empty", () => {
    const { inputEmail, btnSubmit } = setUp();
    fireEvent.change(inputEmail, { target: { value: "test_email@gmail.com" } });
    expect(btnSubmit).toBeDisabled();
  });

  it("Button Login/Register enabled when email not empty and password not empty", () => {
    const { inputEmail, inputPassword, btnSubmit } = setUp();
    fireEvent.change(inputEmail, { target: { value: "test_email@gmail.com" } });
    fireEvent.change(inputPassword, { target: { value: "test_password" } });
    expect(btnSubmit).not.toBeDisabled();
  });

  it("Button Login/Register should call login function when click",async () => {
    const { form } = setUp();
    form?.dispatchEvent(new Event("submit", { bubbles: true }));
    await flushPromises()
    expect(login).toBeCalledTimes(1);
  });
  
  it('should render correct Login', () => {
    const login = render(<Login />)
    expect(login).toMatchSnapshot();
  });
});
