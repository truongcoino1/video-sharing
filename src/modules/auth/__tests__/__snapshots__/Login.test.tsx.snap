// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render correct Login 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-container"
      >
        <form
          class="login-form flex-col flex lg:flex-row"
        >
          <input
            aria-label="email"
            class="lg:mr-8 h-[40px] px-8 mt-12 lg:mt-0 border rounded-medium border-gray-300"
            name="email"
            placeholder="Email"
            type="email"
            value=""
          />
          <input
            aria-label="password"
            class="lg:mr-8 h-[40px] px-8 border mt-12 lg:mt-0 rounded-medium border-gray-300"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <button
            class="px-8 lg:mt-0 h-[40px] mt-12 rounded-medium flex items-center justify-center opacity-50 cursor-not-allowed bg-dark-200 text-dark-high "
            disabled=""
            type="submit"
          >
            Login / Register
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-container"
    >
      <form
        class="login-form flex-col flex lg:flex-row"
      >
        <input
          aria-label="email"
          class="lg:mr-8 h-[40px] px-8 mt-12 lg:mt-0 border rounded-medium border-gray-300"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
        <input
          aria-label="password"
          class="lg:mr-8 h-[40px] px-8 border mt-12 lg:mt-0 rounded-medium border-gray-300"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
        <button
          class="px-8 lg:mt-0 h-[40px] mt-12 rounded-medium flex items-center justify-center opacity-50 cursor-not-allowed bg-dark-200 text-dark-high "
          disabled=""
          type="submit"
        >
          Login / Register
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
